compliance/collection/nodeinventorizer/caching_inventorizer_test.go

```diff
@@ -0,0 +1,128 @@
+	suite.Run(t, new(TestComplianceCachingSuite))
+}
+
+// run before each test
+// func (s *TestComplianceCachingSuite) SetupTest() {
+//	s.sleeper = mockSleeper{callCount: 0}
+//	s.cs = *NewCachingScanner(
+//		fmt.Sprintf("%s/inventory-cache", s.T().TempDir()),
+//		s.sleeper.mockWaitCallback,
+//	)
+//}
+
+func (s *TestComplianceCachingSuite) TestMin() {
+	initial := 2 * time.Second
+	maxBackoff := 10 * time.Second
+
+	actual := min(initial, maxBackoff)
+
+	// On successful test, the duration must not be overwritten
+	s.Equal(initial, actual)
+}
+
+func (s *TestComplianceCachingSuite) TestMinMaxBackoff() {
+	initial, _ := time.ParseDuration("100h")
+	maxBackoff, _ := time.ParseDuration("10s")
+
+	actual := min(initial, maxBackoff)
+
+	s.Equal(maxBackoff, actual)
+}
+
+func (s *TestComplianceCachingSuite) TestCalcNextBackoff() {
+	initial, _ := time.ParseDuration("2s")
+	cache := initial
+	maxBackoff, _ := time.ParseDuration("10s")
+	cs := *NewCachingScanner(mockScanner{}, "", cache, initial, maxBackoff, func(time.Duration) {})
+	expectedBackoff := initial * backoffMultiplier
+
+	newBackoff := cs.calcNextBackoff(initial)
+
+	s.Equal(expectedBackoff, newBackoff)
+}
+
+func (s *TestComplianceCachingSuite) TestCalcNextBackoffUpperBoundary() {
+	initial, _ := time.ParseDuration("8s")
+	cache := initial
+	maxBackoff, _ := time.ParseDuration("10s")
+	cs := *NewCachingScanner(mockScanner{}, "", cache, initial, maxBackoff, func(time.Duration) {})
+
+	newBackoff := cs.calcNextBackoff(initial)
+
+	s.Equal(maxBackoff, newBackoff)
+}
+
+func (s *TestComplianceCachingSuite) TestTriggerNodeInventoryWithoutResultCache() {
+	initial, _ := time.ParseDuration("8s")
+	cache := initial
+	maxBackoff, _ := time.ParseDuration("10s")
+	inventoryCachePath := fmt.Sprintf("%s/inventory-cache", s.T().TempDir())
+	cs := *NewCachingScanner(mockScanner{}, inventoryCachePath, cache, initial, maxBackoff, func(time.Duration) {})
+	nodeName := "testme"
+
+	actual, e := cs.Scan(nodeName)
+
+	s.NoError(e)
+	s.Equal(nodeName, actual.GetNodeName())
+}
+
+func (s *TestComplianceCachingSuite) TestTriggerNodeInventoryHonorBackoff() {
+	sleeper := mockSleeper{callCount: 0}
+	expectedDuration, _ := time.ParseDuration("4s")
+
+	initial, _ := time.ParseDuration("2s") // must be lower than BackoffDuration in file
+	cache := initial
+	maxBackoff, _ := time.ParseDuration("10s")
+	inventoryCachePath := fmt.Sprintf("%s/inventory-cache", s.T().TempDir())
+	cs := *NewCachingScanner(mockScanner{}, inventoryCachePath, cache, initial, maxBackoff, sleeper.mockWaitCallback)
+
+	w := inventoryWrap{
+		CacheValidUntil:      time.Time{},
+		RetryBackoffDuration: 4000000000, // 4 seconds
+		CachedInventory:      "",
+	}
+	jsonWrap, e := json.Marshal(&w)
+	s.NoError(e)
+	e = os.WriteFile(cs.inventoryCachePath, jsonWrap, 0600)
+	s.NoError(e)
+
+	_, err := cs.Scan("testme")
+
+	s.NoError(err)
+	s.Equal(expectedDuration, sleeper.receivedDuration)
+	s.Equal(1, sleeper.callCount)
+}
```

Comment:
A good way to check if you've done enough for testing is to verify if tests cover all interesting code branches. For this right-click `caching_inventorizer_test.go` in GoLand, select `More Run/Debug` | `Run '...' with Coverage`.
Let it run tests, then check back `caching_inventorizer.go`. Lines with yellow color are where only part of branches is executed. Lines with red are where no coverage at all, i.e. that code does not get executed during tests at all.
Right now, I see many interesting spots but those ideally should be the ones covered by unit tests since unit tests provide the cheapest form of verifying branching and permutations compared to more high-level tests (integration, end-to-end).

![image](https://user-images.githubusercontent.com/537715/222515092-86aa69ed-cd7c-45d6-9f37-70d37773e316.png)

You don't need to craft cases for all just to get to 100% coverage, but the interesting ones like returning cached inventory, skipping expired cache, and error during scan with subsequent retry deserve to be tested, IMO.

Reply:
I also think that adding some more coverage for the code we wrote here would be beneficial. Especially the error handling should be covered.